### PR TITLE
Environment-owned dataset: make `train_dataset` optional

### LINKED ISSUE #5903: RFC: Agent RL training with environment: environment-owned dataset, if any
**Scope:** `GRPOTrainer` + `AsyncGRPOTrainer`. Same pattern applies to any online trainers with environment support.
**Out of scope:** reward de-correlation (separate RFC).

## Summary

Let the environment own the dataset, if any.
Today the trainer owns the data. The trainer samples a prompt from the dataset, resets the environment with that prompt as argument.
This RFC shifts that ownership to the env, which can hold its own corpus, self-sample, or generate procedurally.

For some environments, a dataset does not make sense (e.g., procedurally generated environments). This new design should allow this, instead of the current dummy dataset workaround.

## Other data points

- verifiers: the **environment owns the dataset**. But an env **must** have a dataset (it raises if none is provided), which I don't think is a good design choice, see below.
- OpenEnv: most environments don't need any external dataset to sample from.

### Before

Image

**Before:** the Trainer samples from the dataset it owns and pushes the data into the env via
`reset(prompt)`.

### After

Image

**After:** the Trainer only calls `env.reset()` and the env returns the prompt (sampled
from its own dataset, if any).

### Example 1, wordle-like, i.e. when a dataset would fit

For some environments, like Wordle, there is a natural notion of dataset (the list of target words).

#### Before

Two options, both awkward. Either the trainer owns the words and iterates over them, with each row
passed to `reset` as a kwarg:

```python
dataset = load_dataset("my_words")  # rows like {"prompt": "Guess the 5-letter word.", "target": "crane"}

class WordleEnvironment:
    def reset(self, target: str) -> None:    # receives the row's columns
        self._target = target

    def guess(self, word: str) -> str:
        """Submit a guess, get per-letter feedback."""
        ...

trainer = GRPOTrainer(
    model=my_model,
    reward_funcs=my_reward_func,
    train_dataset=dataset,                   # trainer owns and iterates the words
    environment_factory=WordleEnvironment,
)
```

…or the environment owns the words, but you still have to fabricate a dummy dataset whose only job is
to drive the loop length:

```python
dummy = Dataset.from_dict({"prompt": ["Guess the 5-letter word."] * 10_000})  # placeholder rows

class WordleEnvironment:
    def __init__(self):
        self.dataset = load_dataset("my_words")

    def reset(self, **kwargs) -> None:       # ignores the dummy row
        self._target = sample(self.dataset)

    def guess(self, word: str) -> str: ...

trainer = GRPOTrainer(
    model=my_model,
    reward_funcs=my_reward_func,
    train_dataset=dummy,                     # exists only to set the number of steps
    environment_factory=WordleEnvironment,
)
```

#### After

Everything task-related is owned by the environment; no dummy dataset:

```python
class WordleEnvironment:
    def __init__(self):
        self.dataset = load_dataset("my_words")

    def reset(self) -> str:
        self._target = sample(self.dataset)
        return "Guess the 5-letter word."

    def guess(self, word: str) -> str:
        """Submit a guess, get per-letter feedback."""
        ...

trainer = GRPOTrainer(
    model=my_model,
    reward_funcs=my_reward_func,
    environment_factory=WordleEnvironment,   # owns the words
    args=GRPOConfig(max_steps=1000),         # required when no dataset
)
```

### Example 2, chess-like, i.e. when a dataset does not fit

In chess the initial state is always the same starting position (or sampled by self-play). There is
no corpus to iterate — forcing a dataset is purely artificial.

#### Before

You must fabricate a dummy dataset just to drive the training loop:

```python
dummy = Dataset.from_dict({"prompt": [""] * 10_000})  # placeholder, content ignored

class ChessEnvironment:
    def reset(self, **kwargs) -> str:
        self._board = chess.Board()          # always the standard starting position
        return str(self._board)

    def move(self, uci: str) -> str:
        """Play a move in UCI notation, get the opponent's reply."""
        ...

trainer = GRPOTrainer(
    model=my_model,
    reward_funcs=my_reward_func,
    train_dataset=dummy,                     # only there to set the number of steps
    environment_factory=ChessEnvironment,
)
```

#### After

No dataset at all; `max_steps` sets the training length:

```python
class ChessEnvironment:
    def reset(self) -> str:
        self._board = chess.Board()
        return str(self._board)

    def move(self, uci: str) -> str:
        """Play a move in UCI notation, get the opponent's reply."""
        ...

trainer = GRPOTrainer(
    model=my_model,
    reward_funcs=my_reward_func,
    environment_factory=ChessEnvironment,
    args=GRPOConfig(max_steps=1000),         # no dataset → max_steps required
)
```

## What this would imply

`transformers.Trainer` assumes that the trainer has a dataset to determine the number of training steps, and falls back to max_steps when the dataset doesn't have a len. If we make the dataset optional, we should add a check to ensure that `max_steps` is provided when a dataset isn't provided at all.

### Group-state sharing is conventional, not required

> [!IMPORTANT]
> This is not inherent to the new design!

GRPO assumes that we reset the environment to the exact same state for all G members of the group. But with the current and new environment implementation, nothing enforces this.

The only way to do it right now is to have, e.g., a "seed" column in the dataset, and have the env reset to the seed it receives.

Note that technically, we could run GRPO with heterogeneous states across the G members, the baseline is only noisier, and it is not the spirit of GRPO, but it would probably work.

**Recommendation**: always pass a `group_id` to `reset()`, and let the env decide what to do with it. This allows group-state sharing as a convention without enforcing it. Its use would be a good practice to be documented.

**Alternative (not recommended)**: instantiate 1 env, and try to deep-copy it G times at each reset. This is brittle and may not even be possible for some envs.

### PR DESCRIPTION
**Pull request by qgallouedec**
Closes #5903.

## What

When an `environment_factory` is provided, the environment now owns the data. `train_dataset` becomes optional: `reset()` returns the prompt.

This removes the need to fabricate a dummy dataset just to drive the loop (procedural envs like chess, self-sampling envs like Wordle).

Applies to both `GRPOTrainer` and `AsyncGRPOTrainer`.

## Behavior

- No `train_dataset` + `environment_factory` → a placeholder dataset of empty prompts is synthesized internally
- No `train_dataset` + no `environment_factory` → raises (unchanged intent).
- A provided dataset no longer needs a `"prompt"` column: any missing prompt defaults to an empty message that `reset()` supplies. This makes multi-environment **routing** datasets that carry only an `environment` column work (the env owns the prompt).
- Providing a `train_dataset` still works exactly as before: its rows feed `reset(**kwargs)` and its length drives the loop.

## Group-state note

Because the environment self-samples on each `reset()`, the `G` members of a group may not share the same initial state (slightly noisier baseline). Passing a `group_id` to `reset()` for deterministic shared state is left to a follow-up.

---

> [!NOTE]
> **Medium Risk**
> Changes core GRPO/AsyncGRPO data and prompt wiring; behavior is guarded and tested, but env-owned self-sampling can make group baselines noisier when group members get different `reset()` states.
> 
> **Overview**
> **`train_dataset` is optional** for `GRPOTrainer` and `AsyncGRPOTrainer` when a single `environment_factory` is set: the environment supplies tasks via `reset()` (return string = user prompt), and **`max_steps`** defines how long training runs. The trainer synthesizes a small placeholder dataset of empty prompts to drive the dataloader.
> 
> **Validation:** no dataset and no environment still errors; dict multi-env factories still require a routing dataset with an `environment` column; env-owned training without positive `max_steps` errors.
> 
> **Rollout path:** with environments, missing `prompt` on a row defaults to an empty user message before `reset()` observation is applied—so routing datasets can be **`environment`-only**. Without an environment, missing `prompt` still fails.
> 
> **Reward models:** conversational detection for nn.Module reward funcs now uses `is_conversational({"prompt": prompts[0]})` in GRPO, RLOO, and SDPO (fixes prompts-only batches).
> 
> **Docs** reframe env-first training, optional external datasets, and multi-env routing examples; **tests** cover guards and env-owned training without a zen prompt dataset.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9cc8d5c820871725959a21ee472af20bb41444b7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by adithya-s-k — COMMENTED** [review]
Approving the direction from the Harbor / OpenReward integration side — this is the right move, and the backward-compatible path (an external dataset still drives length; `prompt` just becomes optional) drops into both specs without breaking them. Filed #6354 to adopt environment-owned reward + dataset there on the back of this.

Two non-blocking follow-ups from that angle:

**1. Self-sampling + `num_generations > 1` quietly degrades the GRPO baseline.** You flag this in the group-state note, but it's a real footgun: with an external dataset the `G` rollouts of a group share one task, so advantages are computed over the same initial state; with env-owned self-sampling each `reset()` can draw a different task, so the group's baseline is taken across *different* problems. Harbor/OpenReward sidestep it by staying dataset-driven (one task per row → shared across the group), which this PR preserves. For the new path, a `warn` when there's no dataset and `num_generations > 1` (until `group_id` lands) would catch it — it won't show up in the loss curve.

**2. No-dataset + `dict` `environment_factory` (#6002) don't compose.** The placeholder is `{"prompt": ...}` with no `environment` column, so with a routing dict the first batch hits `raise ValueError("Example has environment=None ...")` — confusing, since the user passed no examples. Since env-owned data implies a single owning env, worth rejecting a `dict` factory upfront in the no-dataset branch (or stating single-factory-only in the docs).

Neither blocks — great ergonomics win for procedural / self-sampling envs.

**Review by adithya-s-k — APPROVED** [review]

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -808,8 +812,21 @@ def get_reward(environments, _env_type=env_type, **kwargs):
...
+            if self.environment_factories is None:
+                raise ValueError("`train_dataset` is required unless an `environment_factory` is provided.")
+            if args.max_steps <= 0:
+                raise ValueError(
+                    "When training without a `train_dataset` (the environment owns the data and returns the prompt "
+                    "from `reset()`), `max_steps` must be set to a positive value to define the training length. Set "
+                    "it via `GRPOConfig(max_steps=...)`."
+                )
+            num_placeholder_rows = args.generation_batch_size // args.num_generations
+            train_dataset = Dataset.from_dict({"prompt": [[{"role": "user", "content": ""}]] * num_placeholder_rows})
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:860]: false positive. `Dataset.from_dict` copies rows through Arrow, so `ds[i]` returns an independent object each access

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -2130,6 +2147,12 @@ def _generate_and_score_completions(
         device = self.accelerator.device
         mode = "train" if self.model.training else "eval"
 
+        # `prompt` is optional when the environment owns the data (e.g. a multi-environment routing dataset that carries
+        # only an `environment` column, or no dataset at all); each rollout's `reset()` then supplies it. Default it on
+        # the rows so every prompt-derived check downstream (conversational detection, multimodal handling) stays
+        # consistent.
+        for x in inputs:
+            x.setdefault("prompt", [{"role": "user", "content": ""}])
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: Good catch, fixed. The prompt default is now gated on environment_factories is not None, so a normal dataset with a missing prompt fails fast as before.

File: trl/experimental/async_grpo/async_rollout_worker.py [hunk 3]
```diff
@@ -398,7 +398,8 @@ async def _generate_loop(self, stop_event: asyncio.Event) -> None:
                             {k: v for k, v in row.items() if k != "environment"} if self._multi_environment else row
                         )
                         observation = environment.reset(**reset_kwargs)
-                    prompt = row["prompt"]
+                    # `prompt` is optional when the environment owns the data; `reset()` then supplies it.
+                    prompt = row["prompt"] if "prompt" in row else [{"role": "user", "content": ""}]
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/async_grpo/async_rollout_worker.py]: Same fix as the sync path: the default now only applies when an environment_factory is set; otherwise a missing prompt still raises.

**Comment by qgallouedec** [comment]
Thanks! On the two follow-ups:

- Agreed it's a footgun, kept it in the docs note for now; the runtime warning would false-alarm on deterministic envs (e.g. chess), so I'll fold it in with `group_id`.
- Fixed: a dict factory without a dataset now raises upfront (routing needs the environment column) instead of the confusing `environment=None` error. We can consider supporting this case though: randomly sample an environment? let's see depending on the real use cases

File: tests/test_grpo_trainer.py [hunk 4]
```diff
@@ -3061,13 +3050,16 @@ def format_reward(completions, **kwargs):
             return [0.0 for _ in completions]
 
         training_args = GRPOConfig(
-            output_dir=self.tmp_dir, per_device_train_batch_size=2, num_generations=2, report_to="none"
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply tests/test_grpo_trainer.py:3218]: Fixed, this test uses a dict factory, so it now gets a routing dataset with an environment column

[the remark this answers is not in the document]
Comment (qgallouedec) [reply tests/test_grpo_trainer.py:3218]: fixed

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -2132,7 +2155,15 @@ def _generate_and_score_completions(
         device = self.accelerator.device
         mode = "train" if self.model.training else "eval"
 
-        prompts = [x["prompt"] for x in inputs]
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:2244]: Fixed in b953e38, conversationality is now detected from the actual prompt (is_conversational({"prompt": prompts[0]})) instead of the raw batch row, which lacks a prompt column in the environment-owned case. Applied to GRPO, RLOO, and sDPO for consistency.

File: trl/experimental/async_grpo/async_rollout_worker.py [hunk 6]
```diff
@@ -500,7 +500,12 @@ async def _generate_loop(self, stop_event: asyncio.Event) -> None:
                             {k: v for k, v in row.items() if k != "environment"} if self._multi_environment else row
                         )
                         observation = environment.reset(**reset_kwargs)
-                    prompt = row["prompt"]
+                    # `prompt` is optional only when an environment owns the data; `reset()` then supplies it. Without
+                    # an environment, a missing `prompt` is a malformed dataset and must still fail fast (KeyError).
+                    if "prompt" not in row and self.environment_factories is not None:
```
Comment (AmineDiro) [inline trl/experimental/async_grpo/async_rollout_worker.py:505]: Is this possible ? because we set placeholder dataset :

```python
   train_dataset = Dataset.from_dict({"prompt": [[{"role": "user", "content": ""}]] * num_placeholder_rows})
```

File: trl/experimental/async_grpo/async_grpo_trainer.py [hunk 7]
```diff
@@ -568,6 +572,27 @@ def __init__(
...
+            # which cycles it indefinitely (so its length only needs to be non-degenerate).
+            if environment_factory is None:
+                raise ValueError("`train_dataset` is required unless an `environment_factory` is provided.")
+            if isinstance(environment_factory, dict):
+                raise ValueError(
+                    "A `dict` `environment_factory` (multiple environments) requires a `train_dataset` with an "
+                    "`environment` column to route each example to its environment. Provide a dataset, or pass a "
+                    "single environment factory."
+                )
+            if self.args.max_steps <= 0:
```
Comment (AmineDiro) [inline trl/experimental/async_grpo/async_grpo_trainer.py:615]: I wonder how this will play out with 🤔 
https://github.com/huggingface/trl/pull/6385

**Review by AmineDiro — APPROVED** [review]
All good, just wanted to point out the interaction with #6385  for AsyncGRPOTrainer

File: trl/experimental/async_grpo/async_rollout_worker.py (hunk 6 above)
Comment (qgallouedec) [reply trl/experimental/async_grpo/async_rollout_worker.py:505]: Yes; the placeholder isn't what this branch is for. It covers "no dataset at all", and those rows always have a prompt. This handles the other case: a user-provided dataset + environment_factory with no prompt column, e.g. a routing dataset carrying only environment. The environment_factories is not None half keeps a genuinely malformed dataset failing fast. Same shape as GRPOTrainer after 6013034f.

File: trl/experimental/async_grpo/async_grpo_trainer.py (hunk 7 above)
Comment (qgallouedec) [reply trl/experimental/async_grpo/async_grpo_trainer.py:615]: right, they compose: #6385 says an explicit max_steps > 0 disables the epoch callback, and the env-owned path mandates max_steps > 0. That's this guard. So the callback never engages here.
Good thing too: #6385's target is ceil(num_train_epochs × |dataset|), and our |dataset| is just placeholder rows sized to one generation round, so that target would be meaningless. #6385 should add a test pinning "no dataset ⇒ epoch callback inactive".
